apis/v1alpha2/httproute_types.go

```diff
@@ -70,20 +70,30 @@ type HTTPRouteSpec struct {
 	// HTTPRoute rules. If no hostname is specified, traffic is routed
 	// based on the HTTPRouteRules.
 	//
-	// Hostname can be "precise" which is a domain name without the terminating
-	// dot of a network host (e.g. "foo.example.com") or "wildcard", which is
-	// a domain name prefixed with a single wildcard label (e.g. `*.example.com`).
-	// The wildcard character `*` must appear by itself as the first DNS
-	// label and matches only a single label.
-	// You cannot have a wildcard label by itself (e.g. Host == `*`).
 	// Requests will be matched against the Host field in the following order:
 	//
-	// 1. If Host is precise, the request matches this rule if
+	// 1. If Hostname is precise, the request matches this rule if
 	//    the HTTP Host header is equal to Host.
-	// 2. If Host is a wildcard, then the request matches this rule if
+	// 2. If Hostname is a wildcard, then the request matches this rule if
 	//    the HTTP Host header is to equal to the suffix
 	//    (removing the first label) of the wildcard rule.
 	//
+	// If a hostname is specified by the Listener that the HTTPRoute is bound
+	// to, all hostnames specified here must match specified hostname as per
+	// the rules above. If any hostname does not match, then the entire Route
+	// is not admitted, and the implementation must raise an 'Admitted'
+	// Condition with a status of `false` for that Listener.
+	//
+	// Examples:
+	// - A Listener with unspecified, empty, or `*` values for Hostname matches
+	//   any HTTPRoute hostname.
+	// - A Listener with `test.foo.com` as the hostname matches *only*
```

Comment:
IMHO no, because it's a 2-level match. A request would have to match the listener (and implicitly the TLS certificates attached to the listener) before it can match a hostname on any routes.

Reply:
There is a distinction between the runtime matching and configuration matching.

If the listener says test.foo.com, I would agree at *runtime* we can only match test.foo.com. But I don't see why I cannot have a route that specifies `*.foo.com` be used.

For example, I create one route with `*.foo.com`, then select a few gateways with hostnames `a.foo.com`, `b.foo.com`, etc. 

This makes it consistent with the other direction as well - the listener and route hostname are a [bidirectional wildcard](https://github.com/istio/istio/blob/65bd2ac632cbd6ec8e3fd85bb9874f685aea0cfd/pkg/config/host/name.go#L36)

Reply:
> There is a distinction between the runtime matching and configuration matching.
> 
> If the listener says test.foo.com, I would agree at _runtime_ we can only match test.foo.com. But I don't see why I cannot have a route that specifies `*.foo.com` be used.

Oh I see what you mean. Yeh that seems perfectly reasonable to me. The runtime effect would be that only `test.foo.com` is exposed.

Reply:
So, you're thinking of the usecase where you have a service, and you want to be able to serve the site out of a bunch of domains? I can kind of see that, I guess.

But our experience when implementing the Ingress API (which does technically allow this pattern) was that it created a very large number of edge cases, especially because there's a *list* of hostnames. But that was because of the interaction between the hostnames under the route section and the hostnames under the TLS section.

I think the Istio matching code makes how the matches would work pretty clear.

tl;dr I'm okay with changing this if everyone else is.

Reply:
> So, you're thinking of the usecase where you have a service, and you want to be able to serve the site out of a bunch of domains? I can kind of see that, I guess.

TBH that example is probably a bit contrived, a more common one is the route has `*` ("I don't care how this is exposed") and the gateway picks the domain

Reply:
I can see that one too, thanks.